sdk/release.sh

```diff
@@ -17,60 +17,114 @@ uhoh() {
-SNAPSHOT_REGEX="^${STABLE_REGEX}-(snapshot|adhoc)\.\d{8}\.\d+(\.\d+)?\.v[0-9a-f]{8}$"
+SNAPSHOT_REGEX="^${STABLE_REGEX}-snapshot\.\d{8}\.\d+(\.\d+)?\.v[0-9a-f]{8}$"
+ADHOC_REGEX="^${STABLE_REGEX}-adhoc\.\d{8}\.\d+(\.\d+)?\.v[0-9a-f]{8}$"
 RC_REGEX="^${STABLE_REGEX}-rc\d+$"
-VERSION_REGEX="(^$STABLE_REGEX$)|($SNAPSHOT_REGEX)|($RC_REGEX)"
+VERSION_REGEX="(^$STABLE_REGEX$)|($SNAPSHOT_REGEX)|($ADHOC_REGEX)|($RC_REGEX)"
 
 function file_ends_with_newline() {
     [[ $(tail -c1 "$1" | wc -l) -gt 0 ]]
 }
 
+commit_belongs_to_branch() {
+    git branch --all --format='%(refname:short)' --contains="$1" \
+      | grep -F -x "$2" >/dev/null 2>/dev/null # grep -q flag quits early and causes a SIGPIPE signal in git branch
+}
+
 check() {
     local sha ver ver_sha
     if ! file_ends_with_newline LATEST; then
         echo "LATEST file does not end with newline. Please correct."
         exit 1
     fi
-    while read line; do
+    while read -r line; do
         sha=$(echo "$line" | gawk '{print $1}')
         ver=$(echo "$line" | gawk '{print $2}')
         split=$(echo "$line" | gawk '{print $3}')
-        if ! echo "$ver" | grep -q -P $VERSION_REGEX; then
+        if ! echo "$ver" | grep -q -P "$VERSION_REGEX"; then
             echo "Invalid version number in LATEST file, needs manual correction."
             echo "Offending version: '$ver'."
             exit 1
         fi
-        if is_snapshot $ver; then
-            ver_sha=$(echo $ver | sed 's/.*\.v//')
+        if is_snapshot_or_adhoc "$ver"; then
+            ver_sha=${ver/*.v/}
             if ! [ "${sha:0:8}" = "$ver_sha" ]; then
                 echo "$ver does not match $sha, please correct. ($ver_sha != ${sha:0:8})"
                 exit 1
             fi
         fi
 
+        version_without_trailer=$(echo "$ver" | grep -oP "^$STABLE_REGEX")
+        version_arr=(${version_without_trailer//./ })
+        if is_stable_or_rc "$ver"; then
+            target_branch="origin/release/${version_arr[0]}.${version_arr[1]}.x"
+            if missing_target_branch "$version_without_trailer"; then
+                echo "Skipping check for release '$line' because version $version_without_trailer has a target branch '$target_branch' which does not exist anymore."
+            elif ! commit_belongs_to_branch "$sha" "$target_branch"; then
+                echo "Error while checking '$line' is a valid release:"
+                echo "Stable or RC release with version $version_without_trailer must come from branch '$target_branch', but commit $sha does not belong to that branch."
+                echo "Stable releases or release candidates must come from their release branch."
+                echo "Commit $sha belongs to the following release branches:"
+                find_release_branches_for_commit "$sha"
+                exit 1
+            else
+                echo "Commit $sha belongs to branch '$target_branch', release '$line' is valid"
+            fi
+        elif is_snapshot "$ver"; then
```

Comment:
The PR description says we allow snapshot for branches that don't exist yet but I don't see that logic here. (I see it for stable versions, for branches that don't exist anymore above.)

Reply:
From what I see, if you have a snapshot version higher than the highest release line, it just checks its from main-2.x/main respectively. (in `release_branch_for_version`) Did we want something more permissive?

Reply:
We allow commits for branches that don't exist yet by allowing main/main-2.x to originate versions that are higher than any other version. For example, if there is a release/3.0.x and release/3.2.x branch available, then main can *only* originate 3.3.x. In that scenario, do you think we should also allow release/3.1.x? I opted against it because it seemed illogical.